### PR TITLE
Update API to enable Simulated Annealing on NEBM model

### PR DESCRIPTION
**Pull request by AlessandroPierro**
Issue Number: #174 

Objective of pull request:

## Pull request checklist

Your PR fulfills the following requirements:
- [x] [Issue](https://github.com/lava-nc/lava-optimization/issues) created that explains the change and why it's needed
- [x] Tests are part of the PR (for bug fixes / features)
- [x] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [x] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`pyb`) passes locally
- [x] Build tests (`pyb -E unit`) or (`python -m unittest`) passes locally

## Pull request type
Please check your PR type:
- [x] Feature
- [x] Refactoring (no functional changes, no api changes)

## What is the current behavior?

-

## What is the new behavior?

-

## Does this introduce a breaking change?

- [x] Yes

### CODE REVIEW

File: src/lava/lib/optimization/solvers/generic/hierarchical_processes.py [hunk 1]
```diff
@@ -357,3 +357,118 @@ def __init__(
...
+    """
+
+    def __init__(
+            self,
+            *,
+            max_temperature: int = 10,
+            min_temperature: int = 0,
+            delta_temperature: int = 1,
+            steps_per_temperature: int = 100,
+            refract_scaling: int = 2,
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/hierarchical_processes.py:393]: The refractory values should be in the range ~ 0-20. Thus, we need refract_scaling to be rather 
refract_scaling = 14
?

File: src/lava/lib/optimization/solvers/generic/hierarchical_processes.py [hunk 2]
```diff
@@ -357,3 +357,118 @@ def __init__(
...
+            refract: ty.Optional[ty.Union[int, npty.NDArray]],
+            shape: ty.Tuple[int, ...] = (1,),
+            init_state: npt.ArrayLike = 0,
+            input_duration: npt.ArrayLike = 6,
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/hierarchical_processes.py:397]: What is input_duration, and is it used anywhere by SA?

File: src/lava/lib/optimization/solvers/generic/hierarchical_processes.py [hunk 3]
```diff
@@ -357,3 +357,118 @@ def __init__(
...
+        ----------
+        shape: tuple
+            The shape of the set of dynamical systems to be created.
+        init_state: npt.ArrayLike, optional
+            The starting value of the state variable.
+        temperature: npt.ArrayLike, optional
+            the temperature of the systems, defining the level of noise.
+        input_duration: npt.ArrayLike, optional
+            Number of timesteps by which each input should be preserved.
+        min_state: npt.ArrayLike, optional
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/hierarchical_processes.py:417]: Could you elaborate what min_state is, and why it is used by SA?

File: src/lava/lib/optimization/solvers/generic/hierarchical_processes.py [hunk 4]
```diff
@@ -357,3 +357,118 @@ def __init__(
...
+            The minimum value for the state variable. The state variable will be
+            truncated at this value if updating results in a lower value.
+        min_integration: npt.ArrayLike, optional
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/hierarchical_processes.py:420]: Could you elaborate what this parameter does?

File: src/lava/lib/optimization/solvers/generic/nebm/models.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+from lava.magma.core.model.py.type import LavaPyType
+from lava.magma.core.resources import CPU
+from lava.magma.core.sync.protocols.loihi_protocol import LoihiProtocol
+
+
+@implements(proc=NEBM, protocol=LoihiProtocol)
+@requires(CPU)
+@tag('fixed_pt')
+class NEBMPyModel(PyLoihiProcessModel):
+    """Fixed point implementation of Stochastic Constraint Integrate and
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/nebm/models.py:16]: Change docstring:
Fixed point implementation of Boltzmann neuron for solving QUBO problems

File: src/lava/lib/optimization/solvers/generic/nebm/models.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        Fire (SCIF) neuron for solving QUBO problems.
+    """
+    a_in = LavaPyType(PyInPort.VEC_DENSE, int, precision=8)
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/nebm/models.py:19]: a_in, s_sig_out, and s_wta_out should all be 24bit precision. Just change the numbers here. They don't have any functional meaning anyway.

File: src/lava/lib/optimization/solvers/generic/nebm/models.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+    # The local cost
+    state: np.ndarray = LavaPyType(np.ndarray, int, precision=24)
+    spk_hist: np.ndarray = LavaPyType(np.ndarray, int, precision=8)
+
+    temperature: np.ndarray = LavaPyType(np.ndarray, int, precision=8)
+    refract: np.ndarray = LavaPyType(np.ndarray, int, precision=8)
+
+    def __init__(self, proc_params):
+        super(NEBMPyModel, self).__init__(proc_params)
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/nebm/models.py:31]: In Python 3, you can just call
`super().__init__(proc_params)`
without providing arguments to super.

File: src/lava/lib/optimization/solvers/generic/nebm/process.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+class NEBMSimulatedAnnealing(AbstractProcess):
+    """
+    Non-equilibrium Boltzmann (NEBM) neuron model to solve QUBO problems.
+    """
+
+    def __init__(
+        self,
+        *,
+        shape: ty.Tuple[int, ...],
+        max_temperature: int,
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/nebm/process.py:72]: No need to pass max_temperature here.
This value is only used to define the initial value of the temperature Var.

File: src/lava/lib/optimization/solvers/generic/nebm/process.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        ----------
+        shape: Tuple
+            Number of neurons. Default is (1,).
+
+
+        temperature: ArrayLike
+            Temperature of the system, defining the level of noise.
+        refract : ArrayLike
+            Refractory period for each neuron. This is the time for which a
+            neuron 'stays ON'.
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/nebm/process.py:94]: This is the minimum time between two state switches of a neuron.

The refract period also applies before a neuron can be switched off again.

File: src/lava/lib/optimization/solvers/generic/nebm/process.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        init_value : ArrayLike
+            The spiking history with which the network is initialized
+        init_state : ArrayLike
+            The state of neurons with which the network is initialized
+        """
+        super().__init__(
+            shape=shape,
+            max_temperature=max_temperature,
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/nebm/process.py:102]: No need to pass the max_temperature here. It should only be used to initiate the temperature Var

File: src/lava/lib/optimization/solvers/generic/sub_process_models.py [hunk 11]
```diff
@@ -89,6 +88,29 @@ def __init__(self, proc):
                                            noise_precision=noise_precision,
                                            sustained_on_tau=on_tau,
                                            cost_diagonal=diagonal)
+        elif neuron_model == 'nebm-sa':
+            max_temperature = proc.hyperparameters.get("max_temperature", 10)
+            min_temperature = proc.hyperparameters.get("min_temperature", 0)
+            delta_temperature = proc.hyperparameters.get("delta_temperature", 1)
+            steps_per_temperature = proc.hyperparameters.get(
+                "steps_per_temperature", 100)
+            refract_scaling = proc.hyperparameters.get("refract_scaling", 2)
```
Comment (phstratmann) [inline src/lava/lib/optimization/solvers/generic/sub_process_models.py:97]: Also here, I would set refract_scaling default to 14.

**Review by phstratmann — APPROVED** [review]
Only minor changes. Please address them and then feel free to merge.

Great work!

File: src/lava/lib/optimization/solvers/generic/hierarchical_processes.py (hunk 2 above)
Comment (AlessandroPierro) [reply src/lava/lib/optimization/solvers/generic/hierarchical_processes.py:397]: It's not used, probably just an outdated parameter from SCIF/NEBM.

File: src/lava/lib/optimization/solvers/generic/hierarchical_processes.py (hunk 3 above)
Comment (AlessandroPierro) [reply src/lava/lib/optimization/solvers/generic/hierarchical_processes.py:417]: It's not used, probably just an outdated parameter from SCIF/NEBM.

File: src/lava/lib/optimization/solvers/generic/hierarchical_processes.py (hunk 4 above)
Comment (AlessandroPierro) [reply src/lava/lib/optimization/solvers/generic/hierarchical_processes.py:420]: It's not used, probably just an outdated parameter from SCIF/NEBM.

**Pull request merged**